Added Test For Glossaries

AhmedLeithy opened the pull request:
I have added a test function for glossaries similar to the one implemented in [the python sdk](https://github.com/Azure/azure-sdk-for-python/pull/19931). 

This is still a draft as I am facing an issue where the test constantly "..failed playback, but was successfully re-recorded (it should pass if re-run)." This may be due to the fact that Requests sent by the test do not match the recorded session json (suggested by @maririos).

Feedback would be greatly appreciated !

AhmedLeithy committed:
Added Translation Test Using Glossary

AhmedLeithy committed:
Made the test Live Only.

AhmedLeithy commented:
After Debugging and comparing session records, I believe I have arrived at what is causing this issue:

The request starting the Translation operation also takes a TranslationGlossary object, which contains the glossarySasURL. Since that URL is variable, I believe that each translation request becomes unique enough for a mismatch. If any reviewer can correct/confirm whether this is enough for a mismatch, I would greatly appreciate it.

The work around would be to have a constant SAS URL for the glossary, which seems undesirable. So I have set the test to LiveOnly, until further review.

maririos reviewed (commented):
The problem that I see is that you are making the `containerName` global for the class instead of local.
When test A starts executing, it will call the `CreateSourceContainerAsync` function and will generate a `sourceContainerName`. Then test B starts and creates a new `sourceContainerName`.
When the glossary test is executed and you do ` var containerClient = GetBlobContainerClient(sourceContainerName);` you could be getting the name of the container for test A or B or who knows.
So because it varies, when you run in playback the chances that values are different are high and that is why the framework needs to re-record.

You could move the setup logic (create glossary container,) to the `DocumentTranslationLiveTestBase` and use it in the `DocumentTranslationWithGlossary` test the container. Similar to how it works for source and target containers

The glossary file could live in the test and be passed to the function that creates the glossary container.

Let me know if that helps

AhmedLeithy commented:
I understand what you mean, but if that was the case, wouldn't the test fail because I will face the same problem with `targetContainerName`? If I end up using test B's `targetContainerName`, I would not find the expected document with the output I'm looking for. But the test constantly succeeds so I'm unsure.

I implemented the solution you suggested, but unfortunately I have been unable to test it due to the `Cannot access source document location with the current permissions.` problem. I will try it again when the issue is resolved.

AhmedLeithy commented:
I understand what you mean. Would the test fail because I will face the same problem with targetContainerName? If I end up using test B's targetContainerName, I would not find the expected document with the output I'm looking for. But the test constantly succeeds so I'm unsure whether this is the reason there is a mismatch with the session records.

I implemented the solution you suggested in a separate branch in  [this commit](https://github.com/AhmedLeithy/azure-sdk-for-net/commit/d64a2d45108705d33b9a8bdae0184334f5f8caa0), but I am still facing the same problem.

I checked the `TestRecordingMismatchException` that is being thrown and this is the exception text:

`"Unable to find a record for the request POST https://document-translation-test.cognitiveservices.azure.com/translator/text/batch/v1.0/batches \r\n Header differences: \r\n Body differences:\r\nRequest and response bodies do not match at index 251:\r\n     request: \"7-30T17%3A18%3A08Z&sr=c&sp=rwl\"\r\n     record:  \"7-30T17%3A08%3A09Z&sr=c&sp=rwl\"\r\n"`

It is not very easy to read but I was able to gather the following by examining the request in the Session Records:
- The mismatch happens in a request performed by `client.StartTranslationAsync(input);` , but we already knew that.
- There is a date/time in the glossarySasUri that changes every time.
- Since the glossarySasUri is in the Request Body for the translation request, there is a mismatch in the request body + content length.

If my understanding is correct, this confirms what I thought in the previous comment about the changing glossarySasUri being the culprit since it is in the request body for the start translation request, which would mean that it has to be a LiveOnly test.

That being said, the issue you mentioned could be problematic, so I will implement changes to avoid those.

maririos commented:
>I implemented the solution you suggested in a separate branch in this commit, but I am still facing the same problem.

Changes here look way better!!! I def like this approach more :)

I do see the problem now, and I am so glad we caught it. As you said, the problem is with the date in the SAS URI. a SAS URI is a secret, so we should not record its value because then that secret becomes available for anyone that sees our recording files and they could use it! and we def don't want to do that.

So to avoid that, we need to tell our recording framework to sanitize that value. We already do that with source and target, but not with glossary. 

Here is an example: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/translation/Azure.AI.Translation.Document/tests/Infrastructure/DocumentTranslationRecordedTestSanitizer.cs#L13

Here is the documentation about how to sanitize something: https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/core/Azure.Core.TestFramework#sanitizing

AhmedLeithy commented:
Okaaaay, that makes sense! I will check those out and implement the changes as soon as possible.

AhmedLeithy committed:
Merge branch 'Azure:main' into TestingGlossaries

AhmedLeithy committed:
Merge branch 'Azure:main' into TestingGlossaries

AhmedLeithy committed:
Cleaning Up Glossary Creation

AhmedLeithy committed:
Merge branch 'TestingGlossaries' of https://github.com/AhmedLeithy/azure-sdk-for-net into TestingGlossaries

AhmedLeithy committed:
Sanitized glossaryURL --> SessionRecords can now be matched

AhmedLeithy committed:
Removed global "targetContainerName" in case of race conditions.

AhmedLeithy committed:
Cleaning Up Code:
CreateGlossaryContainerAsync --> CreateGlossaryAsync
Since we do not use the container sasURi for anything, it makes sense to specialize the function to return a single glossary SasUri.

maririos commented on sdk/translation/Azure.AI.Translation.Document/tests/TranslationOperationLiveTests.cs:
```diff
@@ -431,8 +431,7 @@ public async Task DocumentTranslationWithGlossary()
 
-            var glossaryContainerSasUri = await CreateGlossaryContainerAsync(new List<TestDocument> {new TestDocument (glossaryName, glossaryContent)});
-            Uri glossarySasUri = new Uri(String.Format("{0}{1}{2}{3}/{4}{5}", glossaryContainerSasUri.Scheme, Uri.SchemeDelimiter, glossaryContainerSasUri.Authority, glossaryContainerSasUri.AbsolutePath,glossaryName,glossaryContainerSasUri.Query));
+            var glossarySasUri = await CreateGlossaryAsync(new TestDocument (glossaryName, glossaryContent));
```
love this new update where you moved the generation of the SAS to the `CreateGlossaryAsync` function

maririos commented on sdk/translation/Azure.AI.Translation.Document/tests/Infrastructure/DocumentTranslationLiveTestBase.cs:
```diff
@@ -93,7 +93,38 @@ public async Task<Uri> CreateTargetContainerAsync(List<TestDocument> documents =
+            }
+
+            var expiresOn = DateTimeOffset.UtcNow.AddHours(1);
+            var glossaryContainerSasUri = containerClient.GenerateSasUri(BlobContainerSasPermissions.Read | BlobContainerSasPermissions.Write, expiresOn);
```
curious, looking at other functions like for example `CreateTargetContainerAsync` we just do `containerClient.GenerateSasUri()...` and return that value. why the need to create a new URI here?

AhmedLeithy replied:
Oh, because the glossaryContainerSasUri is used in the following line to build the glossarySasUri, which ends up being the return value of the function.

maririos replied:
what id we just do:
`return containerClient.GenerateSasUri(BlobContainerSasPermissions.Read | BlobContainerSasPermissions.Write, expiresOn);`

AhmedLeithy replied:
Then that would return the SasUri for the container itself, not the SasUri for the "blob", which is the glossary in this case. Other functions like CreateTargetContainer etc. should return the SasUri for the container, since this what is needed as input for creating DocumentTranslationInput. However, to create a TranslationGlossary object, you need the SasUri for the glossary itself.

maririos replied:
>to create a TranslationGlossary object, you need the SasUri for the glossary itself.

Is that true though? Looking at the documentation, you can pass a container or a file https://docs.microsoft.com/en-us/azure/cognitive-services/translator/document-translation/get-started-with-document-translation?tabs=csharp#create-sas-access-tokens-for-document-translation
so if you just leave it as a container, it should work

AhmedLeithy replied:
oh interesting. I hadn't looked at the documentation for this part specifically, I had been following some sample code.

But yea, it does not work. I had tried this before, and I re-tested now to double check and when I do the change that you mentioned, there is no output in the output file. (A blank output file is written). I checked through the Azure portal to make sure that there is in fact a file in the target container, and there was an empty file.

Is this an issue we need to address?

maririos commented:
/azp run net - translation - tests
